nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/model/impl/DeprecatedProperty.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * (C) Copyright 2017 Nuxeo SA (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Kevin Leturc
+ */
+package org.nuxeo.ecm.core.api.model.impl;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Objects;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.PropertyException;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.ecm.core.api.model.PropertyNotFoundException;
+import org.nuxeo.ecm.core.api.model.PropertyVisitor;
+import org.nuxeo.ecm.core.schema.types.Field;
+import org.nuxeo.ecm.core.schema.types.Type;
+
+/**
+ * Property used to declare property as deprecated.
+ *
+ * @since 9.1
+ */
+public class DeprecatedProperty extends AbstractProperty {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Log log = LogFactory.getLog(DeprecatedProperty.class);
+
+    private final Property deprecated;
+
+    private final Property fallback;
+
+    public DeprecatedProperty(Property deprecated) {
+        this(deprecated, null);
+    }
+
+    public DeprecatedProperty(Property deprecated, Property fallback) {
+        this(deprecated.getParent(), deprecated, fallback);
+    }
+
+    private DeprecatedProperty(Property parent, Property deprecated, Property fallback) {
+        super(parent);
+        this.deprecated = deprecated;
+        this.fallback = fallback;
+    }
+
+    @Override
+    public void internalSetValue(Serializable value) throws PropertyException {
+        StringBuilder msg = newDeprecatedMessage();
+        msg.append("Set value to deprecated property");
+        if (fallback != null) {
+            msg.append(" and to fallback property '").append(fallback.getPath()).append("'");
+        }
+        if (log.isTraceEnabled()) {
+            log.warn(msg, new NuxeoException());
```

Comment:
I like to use something like `new NuxeoException("debug stack trace")` in these cases so that the trace is clearly labeled as debug.